### PR TITLE
ISSUE #4739 - Custom tickets - Customisable status frontend

### LINKED ISSUE #4739: Custom tickets: show customisable status in frontend
### Description
For extra flexibility we want to provide the users with the ability for setting up their own status in tickets.
For showing the icons and setting behaviour we will have type of status.

### Goals
- [ ] As an admin of a project, I want to be able to freely define what are my custom ticket status for each ticket template
- [ ] As a user, I want to be able to filter completed tickets on custom ticket statuses.
- [ ] As a user, I want to see the custom status to show a matching symbols

### Tasks
- [ ] Update ui for taking the data from the config
- [ ] Not customizable status should work as before

### Notes
Sample of how the data will come
image

### PR DESCRIPTION
**Pull request by The-Daniel**
This fixes #4739

#### Description
- Defining a custom status in a template config now shows it in the front end
<details><summary>Template with custom status</summary>
<p>

```
{
   "_id":"0f624abe-5313-403e-9723-f555cd961b03",
   "modules":[
      {
         "properties":[
            {
               "type":"boolean",
               "default":false,
               "immutable":true,
               "name":"Boolean"
            },
            {
               "type":"text",
               "immutable":true,
               "name":"Text"
            },
            {
               "type":"longText",
               "immutable":true,
               "name":"Long Text"
            },
            {
               "type":"date",
               "immutable":true,
               "name":"Date"
            },
            {
               "type":"number",
               "immutable":true,
               "name":"Number"
            },
            {
               "type":"image",
               "immutable":true,
               "name":"Image"
            },
            {
               "type":"coords",
               "immutable":true,
               "name":"Coordinates"
            },
            {
               "type":"oneOf",
               "immutable":true,
               "values":"jobsAndUsers",
               "name":"One of"
            },
            {
               "type":"manyOf",
               "immutable":true,
               "values":"riskCategories",
               "name":"Many of"
            },
            {
               "name":"Text (Prefilled)",
               "type":"text",
               "deprecated":true,
               "default":"This text input has a default value",
               "immutable":true
            }
         ],
         "name":"Immutable properties"
      },
      {
         "type":"safetibase",
         "properties":[
            {
               "type":"boolean",
               "default":false,
               "name":"Risk Assessed"
            }
         ]
      },
      {
         "properties":[
            {
               "type":"coords",
               "name":"Colour Undefined"
            },
            {
               "type":"coords",
               "color":[
                  255,
                  0,
                  0
               ],
               "name":"Colour Specified (red)"
            },
            {
               "type":"boolean",
               "default":false,
               "name":"Boolean"
            },
            {
               "type":"coords",
               "name":"Colour By Property (Switch)",
               "color":{
                  "property":{
                     "module":"Coloured Pins",
                     "name":"Boolean"
                  },
                  "mapping":[
                     {
                        "default":[
                           245,
                           245,
                           220
                        ]
                     },
                     {
                        "value":true,
                        "color":[
                           160,
                           255,
                           0
                        ]
                     }
                  ]
               }
            },
            {
               "type":"coords",
               "color":{
                  "property":{
                     "name":"Status"
                  },
                  "mapping":[
                     {
                        "default":[
                           188,
                           190,
                           202
                        ]
                     },
                     {
                        "value":"Open",
                        "color":[
                           107,
                           119,
                           140
                        ]
                     },
                     {
                        "value":"In Progress",
                        "color":[
                           113,
                           86,
                           255
                        ]
                     },
                     {
                        "value":"For Approval",
                        "color":[
                           2,
                           136,
                           209
                        ]
                     },
                     {
                        "value":"Closed",
                        "color":[
                           46,
                           125,
                           50
                        ]
                     },
                     {
                        "value":"Void",
                        "color":[
                           0,
                           0,
                           0
                        ]
                     }
                  ]
               },
               "name":"Colour By Property (Status)"
            },
            {
               "type":"text",
               "name":"Colour Textbox"
            },
            {
               "type":"coords",
               "color":{
                  "property":{
                     "module":"Coloured Pins",
                     "name":"Colour Textbox"
                  },
                  "mapping":[
                     {
                        "default":[
                           255,
                           255,
                           255
                        ]
                     },
                     {
                        "value":"black",
                        "color":[
                           0,
                           0,
                           0
                        ]
                     },
                     {
                        "value":"red",
                        "color":[
                           255,
                           0,
                           0
                        ]
                     },
                     {
                        "value":"green",
                        "color":[
                           0,
                           255,
                           0
                        ]
                     },
                     {
                        "value":"blue",
                        "color":[
                           0,
                           0,
                           255
                        ]
                     },
                     {
                        "value":"hotpink",
                        "color":[
                           255,
                           105,
                           180
                        ]
                     }
                  ]
               },
               "name":"Colour By Property (Textbox)"
            },
            {
               "name":"Colour Threshold Test",
               "type":"oneOf",
               "values":[
                  "PRIMARY_CONTRAST",
                  "TERTIARY_LIGHTEST",
                  "TERTIARY_LIGHTER",
                  "BASE_LIGHTEST",
                  "BASE_LIGHTER",
                  "BASE_LIGHT",
                  "ERROR_LIGHTEST",
                  "BLACK"
               ]
            },
            {
               "name":"Colour by Property (Select)",
               "type":"coords",
               "color":{
                  "property":{
                     "module":"Coloured Pins",
                     "name":"Colour Threshold Test"
                  },
                  "mapping":[
                     {
                        "default":[
                           250,
                           229,
                           154
                        ]
                     },
                     {
                        "value":"PRIMARY_CONTRAST",
                        "color":[
                           255,
                           255,
                           255
                        ]
                     },
                     {
                        "value":"TERTIARY_LIGHTEST",
                        "color":[
                           242,
                           246,
                           252
                        ]
                     },
                     {
                        "value":"TERTIARY_LIGHTER",
                        "color":[
                           232,
                           238,
                           248
                        ]
                     },
                     {
                        "value":"BASE_LIGHTEST",
                        "color":[
                           208,
                           217,
                           235
                        ]
                     },
                     {
                        "value":"BASE_LIGHTER",
                        "color":[
                           193,
                           200,
                           213
                        ]
                     },
                     {
                        "value":"BASE_LIGHT",
                        "color":[
                           188,
                           190,
                           202
                        ]
                     },
                     {
                        "value":"ERROR_LIGHTEST",
                        "color":[
                           249,
                           236,
                           236
                        ]
                     },
                     {
                        "value":"BLACK",
                        "color":[
                           0,
                           0,
                           0
                        ]
                     }
                  ]
               }
            }
         ],
         "name":"Coloured Pins"
      },
      {
         "name":"Unique",
         "properties":[
            {
               "type":"oneOf",
               "values":"jobsAndUsers",
               "name":"One of",
               "unique":true
            },
            {
               "type":"text",
               "name":"Text",
               "unique":true
            },
            {
               "type":"number",
               "name":"Number",
               "unique":true
            }
         ]
      },
      {
         "name":"immutable",
         "deprecated":true,
         "properties":[
            {
               "type":"boolean",
               "default":false,
               "name":"Boolean"
            },
            {
               "type":"text",
               "name":"Text"
            },
            {
               "type":"text",
               "default":"This text input has a default value",
               "name":"Text (Prefilled)"
            },
            {
               "type":"longText",
               "name":"Long Text"
            },
            {
               "type":"date",
               "name":"Date"
            },
            {
               "type":"number",
               "name":"Number"
            },
            {
               "type":"image",
               "name":"Image"
            },
            {
               "type":"coords",
               "name":"Coordinates"
            },
            {
               "type":"oneOf",
               "values":"jobsAndUsers",
               "name":"One of"
            },
            {
               "type":"manyOf",
               "values":"riskCategories",
               "name":"Many of"
            }
         ]
      }
   ],
   "properties":[
      {
         "name":"Colour by Property (Priority)",
         "type":"coords",
         "color":{
            "property":{
               "name":"Priority"
            },
            "mapping":[
               {
                  "default":[
                     107,
                     119,
                     140
                  ]
               },
               {
                  "value":"Low",
                  "color":[
                     2,
                     136,
                     209
                  ]
               },
               {
                  "value":"Medium",
                  "color":[
                     237,
                     108,
                     2
                  ]
               },
               {
                  "value":"High",
                  "color":[
                     190,
                     67,
                     67
                  ]
               }
            ]
         }
      },
      {
         "type":"text",
         "name":"readOnlyOnUI",
         "readOnlyOnUI":true
      }
   ],
   "config":{
      "comments":true,
      "issueProperties":true,
      "pin":{
         "color":{
            "property":{
               "module":"safetibase",
               "name":"Risk Assessed"
            },
            "mapping":[
               {
                  "default":[
                     235,
                     0,
                     0
                  ]
               },
               {
                  "value":true,
                  "color":[
                     0,
                     235,
                     0
                  ]
               }
            ]
         }
      },
      "status":{
         "values":[
            {
               "name":"Awaiting triage",
               "type":"open"
            },
            {
               "name":"Triaged (This option is extra long to test for overflow)",
               "type":"open"
            },
            {
               "name":"Active",
               "type":"active"
            },
            {
               "name":"To verify",
               "type":"review"
            },
            {
               "name":"Approved",
               "type":"done"
            },
            {
               "name":"Won't fix",
               "type":"void"
            }
         ],
         "default":"Awaiting triage"
      }
   },
   "code":"DMO",
   "name":"Demo Template"
}
```

</p>
</details> 

#### Test cases

- Can create a ticket with custom status
- Can change status
- Can correctly filter by 'completed'
- Tabular view works as before

### CODE REVIEW

File: frontend/src/v5/store/tickets/tickets.selectors.ts [hunk 1]
```diff
@@ -109,3 +110,10 @@ export const selectTicketsByContainersAndFederations = createSelector(
 		return sortTicketsByCreationDate(tickets);
 	},
 );
+
+export const selectStatusConfigByTemplateId = createSelector(
+	(state) => state,
+	(state, modelId) => modelId,
+	(state, modelId, templateId) => templateId,
+	(storeState, modelId, templateId) => selectTemplateById(storeState, modelId, templateId)?.config?.status || DEFAULT_STATUS_CONFIG,
+);
```
Comment (sanmont3drepo) [inline frontend/src/v5/store/tickets/tickets.selectors.ts:119]: for reusing  selectTemplateById you can do this:
```suggestion
export const selectStatusConfigByTemplateId = createSelector(
	selectTemplateById,
	(template) => template?.config?.status || DEFAULT_STATUS_CONFIG,
);
```

**Review by sanmont3drepo — CHANGES_REQUESTED** [review]

File: frontend/src/v5/store/tickets/card/ticketsCard.selectors.ts [hunk 2]
```diff
@@ -20,11 +20,11 @@ import { SequencingProperties, TicketBaseKeys, TicketsCardViews } from '@/v5/ui/
 import { createSelector } from 'reselect';
 import { selectTemplateById, selectTemplates, selectTicketById, selectTickets } from '../tickets.selectors';
 import { ITicketsCardState } from './ticketsCard.redux';
-import { getTicketIsCompleted } from './ticketsCard.helpers';
 import { DEFAULT_PIN, getPinColorHex, formatPin } from '@/v5/ui/routes/viewer/tickets/ticketsForm/properties/coordsProperty/coordsProperty.helpers';
```
Comment (Amantini1997) [inline frontend/src/v5/store/tickets/card/ticketsCard.selectors.ts:23]: I don't think the chip should be truncated here 
image

**Review by Amantini1997 — CHANGES_REQUESTED** [review]

**Review by Amantini1997 — APPROVED** [review]

**Review by sanmont3drepo — APPROVED** [review]

**Review by sanmont3drepo — APPROVED** [review]

**Review by Amantini1997 — APPROVED** [review]

File: frontend/src/v5/store/tickets/card/ticketsCard.selectors.ts [hunk 3]
```diff
@@ -20,11 +20,11 @@ import { SequencingProperties, TicketBaseKeys, TicketsCardViews } from '@/v5/ui/
 import { createSelector } from 'reselect';
 import { selectTemplateById, selectTemplates, selectTicketById, selectTickets } from '../tickets.selectors';
```
[the remark this answers is not in the document]
Comment (carmenfan) [reply frontend/src/v5/store/tickets/card/ticketsCard.selectors.ts:21]: it's a general problem with `oneOf` `manyOf` where we should've checked for duplicates at validation. (created an issue for it too : https://github.com/3drepo/3drepo.io/issues/4796)

Don't worry about it, my aim is to fix it before we release... so we can assume this can never happen.
